### ISSUE TITLE
[Question] ads events, how to know preroll ad is ready to play

### ISSUE THREAD

**Issue by oahmedazab** [kaltura/playkit-android#631]
Usually when playing video without ads i do something like that
```
kalturaPlayer.prepare(config)
```
and set listners on player to know when its ready and can play
```
        kalturaPlayer.addListener(this, PlayerEvent.Type.CAN_PLAY) {
            // content loaded can play now
            kalturaPlayer.play()
        }

        kalturaPlayer.addListener(this, PlayerEvent.Type.ERROR) {
           // show error message
        }
```

but after i integrated IMAPlugin non of these listener called!
i only get two events

```
PlayerEvent.Type.STATE_CHANGED with state LOADING
```
and
```
AdEvent.Type.AD_REQUESTED
```

how do you handle this scenario?

**Comment by oahmedazab** [comment]
@giladna if you can share your opinion that would be great 🙂

**Comment by giladna** [comment]
@oahmedazab 

you have to listen to 

```
 player.addListener(this, AdEvent.contentPauseRequested, event -> {
            log("ADS_PLAYBACK_START");
});
        
player.addListener(this, AdEvent.contentResumeRequested, event -> {
            log("ADS_PLAYBACK_ENDED");
});
```

**Comment by oahmedazab** [comment]
i don't receive contentPauseRequested event in case of preroll ads

**Comment by oahmedazab** [comment]
**Scenario 1 (no ads)**
events received after prepare
stateChanged= BUFFERING
stateChanged= LOADING
stateChanged= READY
PlayerEvent.Type.CAN_PLAY


**Scenario 2 (with ads)**
events received after prepare
AdEvent.Type.AD_REQUESTED

**Comment by giladna** [comment]
@oahmedazab 

When using IMA first we load the ads and only after ad started we load the media to the player 
so you have to register to ads events and player events and do your logic accordingly

what is the error that you face?

did you check this sample?

https://github.com/kaltura/kaltura-player-android-samples/tree/develop/AdvancedSamples/IMASample

see events registration example here:

https://github.com/kaltura/playkit-android-samples/blob/9d917993989d2f9445df272a2455921bddfbee37/FullDemo/playkitdemo/src/main/java/com/kaltura/playkitdemo/MainActivity.java#L916

**Comment by giladna** [comment]
@oahmedazab  content pause requested will be not be fired for preroll since no content to resume yet

**Comment by oahmedazab** [comment]
> what is the error that you face?

i just couldn't find adEvent or playerEvent to notify me that the preroll ad is loaded so i can start the player 
is `AdEvent.Type.AD_REQUESTED` is this enough for me to know that the preroll is loaded?

Thanks

**Comment by giladna** [comment]
you do not have to do anything 
player is taking care of this 

once ad is ready it is playing unless you did not call play

you do not have to start the player after some event just call prepare and player.play()

**Comment by oahmedazab** [comment]
in my case i am not calling `player.play()` directly after `player.prepare()`
i need to do some operations after video is ready then play

**Comment by giladna** [comment]
@oahmedazab do  you still need support or this is ok now?

**Comment by oahmedazab** [comment]
@giladna i am doing a workaround to make it work for me right now, but in the future it would be nice to have such an event

**Comment by giladna** [comment]
@oahmedazab 

for pre roll no event you have to wait for 
you can cal player play if it is ready it will play and if not it will take the time

the prepare is async operation and plat will init IMA
if you want to be sure IMA was initialized you can use the AD_REQUESTED

{code}
      player.addListener(this, AdEvent.adRequested, event -> {
            log.d("adRequested  = " + event.adTagUrl);
        });
{code}

**Comment by oahmedazab** [comment]
@giladna great that's the same approach i am using right now
